### PR TITLE
Proposing Accordion component

### PR DESCRIPTION
**Pull request by epilande**
#### What does this PR do?
Add Accordion component.

#### Where should the reviewer start?
Check out `Accordion`, then `AccordionPanel`, then `Collapsible`.

#### What testing has been done on this PR?
Tested on grommet-docs https://github.com/grommet/grommet-docs

#### How should this be manually tested?
Fire up grommet-docs & test in Accordion component doc page

#### Screenshots
accordion-component

#### Do the grommet docs need to be updated?
Already did. See grommet/grommet-docs#62. 😄 

#### Should this PR be mentioned in the release notes?
Sure.

### CODE REVIEW

File: src/js/components/Accordion.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+const CLASS_ROOT = CSSClassnames.ACCORDION;
+
+export default class Accordion extends Component {
+  render () {
+    const { animate, headline, subHeadline, colorIndex, children } = this.props;
+
+    const classes = classnames(
+      CLASS_ROOT,
+      this.props.className
```
Comment (alansouzati) [inline src/js/components/Accordion.js]: should we move this up to the const list so that we have only one place with `this.props` ?

File: src/js/components/Accordion.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    let subHeadlineMarkup;
+    if (subHeadline) {
+      subHeadlineMarkup = (
+        <Paragraph
+          className={`${CLASS_ROOT}__sub-headline`}
+          size="large"
+          align="center"
+        >
```
Comment (alansouzati) [inline src/js/components/Accordion.js]: can we move this to the line before?

File: src/scss/grommet-core/_objects.accordion.scss [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP
+
+.grommet .#{$grommet-namespace}accordion {
```
Comment (alansouzati) [inline src/scss/grommet-core/_objects.accordion.scss]: I believe we don't need the `.grommet` here. We only need that for layers, eg: Menu drop, Layer...

File: src/js/components/Collapsible.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
+// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP
+
+import React, { Component } from 'react';
+import ReactDOM from 'react-dom';
+import TransitionGroup from 'react-addons-transition-group';
```
Comment (alansouzati) [inline src/js/components/Collapsible.js]: I'm assuming we are using transition group directly only because Animate component was not approved yet. Is this interpretation correct?

**Comment by L0ZZI** [comment]
Looking at the video...My expectation would be that only one item could be expandable at a time. That means that as a user clicks on another item in the accordion that isn't the selected item, the previous selection would automatically collapse. I'd prefer that the user doesn't need to "manage" the content. 

Also, we need to change the icons from a "+" to "tab-next" and rotate it to a down position on expansion.

**Comment by alansouzati** [comment]
Nice job. I've sent some feedbacks in the code, but generally your code structure looks really nice and clean 👏 

This component needs to have support for accessibility. I added support for accessibility for the [Tabs](https://github.com/grommet/grommet/blob/master/src/js/components/Tabs.js) component, which is similar to the Accordion. You need to have aria roles "tablist", "tabpanel" and "tab".

This another good example of accessible Accordions [here](https://labs.ssbbartgroup.com/index.php/ARIA_Accordion)

File: src/scss/grommet-core/_objects.accordion.scss (hunk 3 above)
Comment (epilande) [inline src/scss/grommet-core/_objects.accordion.scss]: This is a specificity problem. 

In hpe.min.css:
```
.grommet p {                                    // specificity level 11
  margin-top: 24px;
  margin-bottom: 24px;
}
```
Which would be more specific than the accordion styles i'm adding.
Anyway I'll remove the custom margin top & bottom & `.grommet` from the styles.

File: src/js/components/Collapsible.js (hunk 4 above)
Comment (epilande) [inline src/js/components/Collapsible.js]: No, it uses the same library (react transition group) as the `Animate` component, but this component has a different purpose. The Collapsible component is used to animate `height`. This is done through JavaScript. It's currently not possible to animate dynamic heights with CSS, e.g. `height: 0;` to `height: 100%;` does not work. I've also tried the `max-height` approach, but the animation/transition is very janky & not reliable depending on the panel content.

File: src/js/components/Accordion.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    const accordionChildren = React.Children.map(children, (child, index) => {
+      return React.cloneElement(child, {
+        id: 'accordion-panel-' + index,
+        animate
+      });
+    });
+
+    return (
+      <Box className={classes} colorIndex={colorIndex}>
+        {content}
```
Comment (ericsoderberghp) [inline src/js/components/Accordion.js]: Why does the headline and sub headline need to be part of the core Accordion component? That seems outside the scope of it to me.

File: src/js/components/AccordionPanel.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    );
+
+    let panelControlIcon = <OpenIcon colorIndex="brand" />;
+    if (this.state.isOpen) {
+      panelControlIcon = <CloseIcon colorIndex="brand" />;
+    }
+
+    return (
+      <ListItem className={classes} direction="column" pad="none">
+        <Box
```
Comment (ericsoderberghp) [inline src/js/components/AccordionPanel.js]: Can this be Header instead of Box?

File: src/js/components/AccordionPanel.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+          direction="row"
+          justify="between"
+          align="center"
+          onClick={this._onClickPanel}
+          responsive={false}
+        >
+          <Heading tag="h2" margin="small">{title}</Heading>
+          {panelControlIcon}
+        </Box>
+        <Box
```
Comment (ericsoderberghp) [inline src/js/components/AccordionPanel.js]: I'd put the Collapsible here without the enclosing Box to allow callers to do things like full width Boxes with textures and background images.

File: src/js/components/AccordionPanel.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
+// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP
+
+import React, { Component, PropTypes } from 'react';
+import classnames from 'classnames';
+import Box from './Box';
+import Heading from './Heading';
+import ListItem from './ListItem';
+import OpenIcon from './icons/base/Add';
```
Comment (ericsoderberghp) [inline src/js/components/AccordionPanel.js]: I'd like to review the use of '+' and '-' as the right metaphors for the accordion controls with Chris. We use + and - for NumberInput where it means something entirely different. I'm wondering if up and down arrow heads would be better.

File: src/js/components/AccordionPanel.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        <Box
+          pad={{horizontal: 'medium', vertical: 'small'}}
+          full="horizontal"
+          direction="row"
+          justify="between"
+          align="center"
+          onClick={this._onClickPanel}
+          responsive={false}
+        >
+          <Heading tag="h2" margin="small">{title}</Heading>
```
Comment (ericsoderberghp) [inline src/js/components/AccordionPanel.js]: I think we should rename the property to `heading` and allow the caller to pass either a string or a node in. That will allow callers to use this with other heading levels.

File: src/js/components/AccordionPanel.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+import ListItem from './ListItem';
+import OpenIcon from './icons/base/Add';
+import CloseIcon from './icons/base/Subtract';
+import Collapsible from './Collapsible';
+
+import CSSClassnames from '../utils/CSSClassnames';
+
+const CLASS_ROOT = CSSClassnames.ACCORDION_PANEL;
+
+export default class AccordionPanel extends Component {
```
Comment (ericsoderberghp) [inline src/js/components/AccordionPanel.js]: Since this is a ListItem inside a List, perhaps it should be called AccordionListItem?

File: src/js/components/Accordion.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+import Box from './Box';
+import Headline from './Headline';
+import List from './List';
+import Paragraph from './Paragraph';
+
+import CSSClassnames from '../utils/CSSClassnames';
+
+const CLASS_ROOT = CSSClassnames.ACCORDION;
+
+export default class Accordion extends Component {
```
Comment (ericsoderberghp) [inline src/js/components/Accordion.js]: Do we even need this component? I'm wondering if callers could do:
```
<List>
  <AccordionListItem />
  <AccordionListItem />
</List>
```

File: src/js/components/AccordionPanel.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+    const classes = classnames(
+      CLASS_ROOT,
+      this.props.className,
+      {
+        [`${CLASS_ROOT}--active`]: this.state.isOpen
+      }
+    );
+
+    let panelControlIcon = <OpenIcon colorIndex="brand" />;
```
Comment (ericsoderberghp) [inline src/js/components/AccordionPanel.js]: Most of our other controls are not brand colored. I think we should review whether these are as well with Chris.

File: src/scss/grommet-core/_objects.accordion-panel.scss [hunk 13]
```diff
@@ -0,0 +1,18 @@
+// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP
+
+.grommet .#{$grommet-namespace}accordion-panel {
+
+  > .#{$grommet-namespace}box--clickable {
```
Comment (ericsoderberghp) [inline src/scss/grommet-core/_objects.accordion-panel.scss]: Can we give the Box a class name specific to the accordion context to avoid the direct child selector and keep the CSS flatter?

File: src/scss/grommet-core/_objects.accordion-panel.scss [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+    user-select: none;
+    -webkit-user-select: none;
+  }
+
+  .#{$grommet-namespace}heading,
+  .#{$grommet-namespace}paragraph {
+    max-width: $size-medium;
```
Comment (ericsoderberghp) [inline src/scss/grommet-core/_objects.accordion-panel.scss]: I would think the max-width would be controlled by the outer context. This would allow accordions to be used in a wider variety of situations.

File: src/js/components/Accordion.js (hunk 11 above)
Comment (alansouzati) [inline src/js/components/Accordion.js]: I believe we need Accordion instead of List. We need to add aria-role="tablist" which is specific for components like Tab and Accordion.

**Comment by epilande** [comment]
@alansouzati @L0ZZI @ericsoderberghp thanks for the feedback! I've updated the PR with your suggestions. Let me know if I missed anything.

accordion-component

File: src/js/components/Accordion.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+  }
+};
+
+Accordion.propTypes = {
+  animate: PropTypes.bool,
+  colorIndex: PropTypes.string
+};
+
+Accordion.defaultProps = {
+  openMulti: false,
```
Comment (ericsoderberghp) [inline src/js/components/Accordion.js]: Missing openMulti in propTypes.

File: src/js/components/Accordion.js [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+const CLASS_ROOT = CSSClassnames.ACCORDION;
+
+export default class Accordion extends Component {
+  constructor(props) {
+    super(props);
+    this._activatePanel = this._activatePanel.bind(this);
+
+    this.state = {
+      activeIndex: props.initialIndex
```
Comment (ericsoderberghp) [inline src/js/components/Accordion.js]: `initialIndex` isn't in propTypes.

File: src/js/components/AccordionPanel.js [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+  constructor (props) {
+    super(props);
+    this._onClickPanel = this._onClickPanel.bind(this);
+    this.state = {
+      isOpen: props.isOpen || false
+    };
+  }
+
+  componentWillReceiveProps (nextProps) {
+    if (this.props.isOpen !== nextProps.isOpen) {
```
Comment (ericsoderberghp) [inline src/js/components/AccordionPanel.js]: `isOpen` isn't in propTypes. I'd suggest `active` as a term, as that would align with other components.

File: src/js/components/Accordion.js [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    const classes = classnames(
+      CLASS_ROOT,
+      className
+    );
+
+    const accordionChildren = React.Children.map(children, (child, index) => {
+      return React.cloneElement(child, {
+        id: 'accordion-panel-' + index,
+        isOpen: !openMulti ? (this.state.activeIndex === index) : null,
+        onTitleClick: () => {
```
Comment (ericsoderberghp) [inline src/js/components/Accordion.js]: I don't see `onTitleClick` in any propTypes. Also, can we align it with 'active' and call it `onActive`?

File: src/js/components/Accordion.js [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+          this._activatePanel(index);
+        },
+        animate
+      });
+    });
+
+    return (
+      <Box role="tablist" className={classes} colorIndex={colorIndex} separator="top">
```
Comment (ericsoderberghp) [inline src/js/components/Accordion.js]: I don't know that all consumers of this will want a top separator. Instead, since this is a Box, can we just transfer all Box.propTypes through?

Comment (ericsoderberghp) [inline src/js/components/Accordion.js]: Although, why do we need both the Box and the List? I wonder if we can have Accordion just be a List and let callers put it inside a Box as needed?

File: src/js/components/AccordionPanel.js [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+      {
+        [`${CLASS_ROOT}--active`]: this.state.isOpen
+      }
+    );
+
+    return (
+      <ListItem className={classes} direction="column" pad="none">
+        <Header
+          role="tab"
+          className={`${CLASS_ROOT}-heading`}
```
Comment (ericsoderberghp) [inline src/js/components/AccordionPanel.js]: I think the should be `${CLASS_ROOT}__header` to align with Grommet BEM patterns.

File: src/scss/grommet-core/_objects.accordion.scss [hunk 21]
```diff
@@ -0,0 +1,8 @@
+// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP
+
+.#{$grommet-namespace}accordion {
+
+  &__sub-headline {
```
Comment (ericsoderberghp) [inline src/scss/grommet-core/_objects.accordion.scss]: I don't think this is needed anymore.

**Comment by RyanCCollins** [comment]
Is the accordian component available in a release?  I am using 0.6.10 and don't seem to have the source for it in the components folder.  Thank you!

**Comment by epilande** [comment]
Hi @RyanCCollins the accordion component isn't part of the 0.6.10 release. It has been merged into the master branch which you can access by setting your grommet version in package.json to 
`"grommet": "https://github.com/grommet/grommet/tarball/stable"`

**Comment by RyanCCollins** [comment]
Hi @epilande Thanks so much!! I tried this, but forked it first.  Looks great!
